scripts/build-tools.sh

```diff
@@ -5,69 +5,112 @@
 {
         cat <<EOFUSAGE
-usage: $0 [-t|-f]
+usage: $0 [-c|-f|-h|-l|-p|-t|-T]
+       [-c] Build ctl
+       [-f] Build fuzzer
+       [-h] Display help
+       [-l] Build logger
+       [-p] Build probes
        [-t] Build test topologies
-       [-f] Build fuzzer"
+       [-T] Build topologies
 EOFUSAGE
 }
 
-build_tools()
+# generate Makefiles
+prepare_build()
 {
-        cd "$SOF_REPO/tools"
-        rm -rf build_tools
-        mkdir build_tools
-        cd build_tools
+        if ! [ -d "$BUILD_TOOLS_DIR" ]
+        then
+                mkdir "$BUILD_TOOLS_DIR"
+        fi
+
+        cd "$BUILD_TOOLS_DIR"
         cmake ..
-        make -j "$NO_PROCESSORS"
-}
 
-build_test()
-{
-        cd "$SOF_REPO/tools/build_tools"
-        make tests -j "$NO_PROCESSORS"
+        if ! [ -d "$BUILD_TOOLS_DIR/fuzzer" ]
+        then
+                mkdir "$BUILD_TOOLS_DIR/fuzzer"
+        fi
+        cd "$BUILD_TOOLS_DIR/fuzzer"
+        cmake ../../fuzzer
 }
 
-build_fuzzer()
+build_tool()
 {
-        cd "$SOF_REPO/tools/build_tools"
-        mkdir build_fuzzer
-        cd build_fuzzer
-        cmake ../../fuzzer
-        make -j "$NO_PROCESSORS"
+        tool=$1
+        make -C "$BUILD_TOOLS_DIR/$tool" -j "$NO_PROCESSORS"
 }
 
 main()
 {
-        local DO_BUILD_TEST DO_BUILD_FUZZER SCRIPT_DIR SOF_REPO
+        local DO_BUILD_CTL DO_BUILD_FUZZER DO_BUILD_LOGGER DO_BUILD_PROBES \
+                DO_BUILD_TEST DO_BUILD_TOPOLOGIES SCRIPT_DIR SOF_REPO
         SCRIPT_DIR=$(cd "$(dirname "$0")" && pwd)
         SOF_REPO=$(dirname "$SCRIPT_DIR")
+        BUILD_TOOLS_DIR="$SOF_REPO"/tools/build_tools
         : ${NO_PROCESSORS:=$(nproc)}
 
-        DO_BUILD_TEST=false
+        if [ $# -eq 0 ]
+        then
+                prepare_build
+                build_tool
+                exit
+        fi
+
+        DO_BUILD_CTL=false
         DO_BUILD_FUZZER=false
-        while getopts "tf" OPTION; do
+        DO_BUILD_LOGGER=false
+        DO_BUILD_PROBES=false
+        DO_BUILD_TEST=false
+        DO_BUILD_TOPOLOGIES=false
+
+        while getopts "cfhlptT" OPTION; do
                 case "$OPTION" in
-                t) DO_BUILD_TEST=true ;;
+                c) DO_BUILD_CTL=true ;;
                 f) DO_BUILD_FUZZER=true ;;
+                l) DO_BUILD_LOGGER=true ;;
+                p) DO_BUILD_PROBES=true ;;
+                t) DO_BUILD_TEST=true ;;
+                T) DO_BUILD_TOPOLOGIES=true ;;
+                h) print_usage; exit 1;;
                 *) print_usage; exit 1;;
                 esac
         done
         shift "$(($OPTIND - 1))"
 
-        build_tools
+        prepare_build
+        if "$DO_BUILD_CTL"
+        then
+                build_tool ctl
+        fi
+
+        if "$DO_BUILD_FUZZER"
+        then
+                build_tool fuzzer
+        fi
+
+        if "$DO_BUILD_LOGGER"
+        then
+                build_tool logger
+        fi
+
+        if "$DO_BUILD_PROBES"
+        then
+                build_tool probes
+        fi
```

Comment:
You can use `eval` and a `for` loop to avoid all this repetition:
```shell
DO_BUILD_ctl=true
DO_BUILD_logger=false
for tool in ctl logger probes ... ; do
    if eval '$DO_BUILD_'$tool; then
        build_tool $tool
    fi
done
```
I'm usually not a fan of `eval` as it increases the quoting complexity[*] when most people (including myself) keep getting a single level of quoting wrong. On the other hand, this is a very simple and minimal use of `eval` and it saves a lot of duplication.

The alternative to `eval` is some `tool -> bool` map/associative array. That's much more change though, up to you.

[*] http://rg1-teaching.mpi-inf.mpg.de/unixffb-ss98/quoting-guide.html

Reply:
this is a good trick

Reply:
```
In ./scripts/build-tools.sh line 76:
                l) DO_BUILD_logger=true ;;
                   ^-------------^ SC2034: DO_BUILD_logger appears unused. Verify use (or export if used externally).
```

shellcheck has similar complaint for other variables. Because we are using eval for dynamically using the variable, the static analysis of shellcheck fails to see usage of these variables.

Reply:
Good to know. It's possible to turn off shellcheck warnings with a magic `# shellcheck ...` comment, or we could use `eval` here too but let's worry about that later when we add shellcheck to `sof` CI (it's already in sof-test). For now I'm just suggesting to run it locally.